[web-animations] add an AnimatableProperty type to encapsulate either a standard or custom property

graouts opened the pull request:
#### ea7b3989b369d140fcf55e9962380d663a79d905
<pre>
[web-animations] add an AnimatableProperty type to encapsulate either a standard or custom property
<a href="https://bugs.webkit.org/show_bug.cgi?id=249498">https://bugs.webkit.org/show_bug.cgi?id=249498</a>

Reviewed by Darin Adler.

The animation code is currently heavily biased towards dealing with &quot;standard&quot; CSS properties, represented
by the CSSPropertyID enum. However, with the work to add interpolation support for custom properties (see
bug 249293) and to support CSS Transitions for custom properties (see bug 249399), we really need to deal
with both custom properties and standard properties everywhere we currently deal with only standard
properties.

In this patch we&apos;re adding a new type, AnimatableProperty, which is an std::variant&lt;CSSPropertyID, AtomString&gt;.
We currently use CSSPropertyID as a value for property lists which are HashSet&lt;&gt; and as keys to HashMap&lt;&gt;,
so we must add the relevant traits to support this use for AnimatableProperty.

* Source/WebCore/animation/WebAnimationTypes.h:
(WTF::DefaultHash&lt;WebCore::AnimatableProperty&gt;::hash):
(WTF::DefaultHash&lt;WebCore::AnimatableProperty&gt;::equal):
(WTF::HashTraits&lt;WebCore::AnimatableProperty&gt;::constructDeletedValue):
(WTF::HashTraits&lt;WebCore::AnimatableProperty&gt;::isDeletedValue):

Canonical link: <a href="https://commits.webkit.org/258040@main">https://commits.webkit.org/258040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0e984bc3b8e6b8618539529ef3a95c1dfa0b3ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100738 "Failed to checkout and rebase branch from PR 7780") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9883 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33780 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/110037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170311 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/104727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10822 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/431 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/93139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107883 "Built successfully") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/106521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/93139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/93139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/3600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/9711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/43838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5378 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2886 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/WebAnimationTypes.h:
```diff
@@ -64,6 +67,33 @@ using AnimationEvents = Vector<Ref<AnimationEventBase>>;
 using CSSAnimationCollection = ListHashSet<RefPtr<CSSAnimation>>;
 
+using AnimatableProperty = std::variant<CSSPropertyID, AtomString>;
+using AnimatablePropertyToTransitionMap = HashMap<AnimatableProperty, RefPtr<CSSTransition>>;
```
Why not `Ref<>` instead of `RefPtr<>`? Can map values be null?

graouts replied:
I'll look into that as a follow-up since this is a broader change in the places where `AnimatablePropertyToTransitionMap` will eventually be used (in replacement of the existing `PropertyToTransitionMap`). I filed https://bugs.webkit.org/show_bug.cgi?id=249530 to track that work.

graouts replied:
There's a patch up for review in https://github.com/WebKit/WebKit/pull/8361.

darinadler commented on Source/WebCore/animation/WebAnimationTypes.h:
```diff
@@ -64,6 +67,33 @@ using AnimationEvents = Vector<Ref<AnimationEventBase>>;
+
+template<> struct DefaultHash<WebCore::AnimatableProperty> {
+    static unsigned hash(const WebCore::AnimatableProperty& key) {
+        return WTF::switchOn(key, [] (WebCore::CSSPropertyID property) { return DefaultHash<WebCore::CSSPropertyID>::hash(property); }
```
We normally write these `switchOn` calls in a more vertical style.

darinadler commented on Source/WebCore/animation/WebAnimationTypes.h:
```diff
@@ -64,6 +67,33 @@ using AnimationEvents = Vector<Ref<AnimationEventBase>>;
+template<> struct DefaultHash<WebCore::AnimatableProperty> {
+    static unsigned hash(const WebCore::AnimatableProperty& key) {
+        return WTF::switchOn(key, [] (WebCore::CSSPropertyID property) { return DefaultHash<WebCore::CSSPropertyID>::hash(property); }
+                                , [] (AtomString string) { return DefaultHash<AtomString>::hash(string); });
```
We should try to use `const AtomString&` here to cut down on reference count churn.

darinadler commented on Source/WebCore/animation/WebAnimationTypes.h:
```diff
@@ -64,6 +67,33 @@ using AnimationEvents = Vector<Ref<AnimationEventBase>>;
 
+namespace WTF {
+
+template<> struct DefaultHash<WebCore::AnimatableProperty> {
```
We should set up default hashing for `std::variant` when all values in the variant can be hashed so this doesn’t have to be done specifically for this variant.

graouts replied:
Yes, that would have made my life easier :)

darinadler commented on Source/WebCore/animation/WebAnimationTypes.h:
```diff
@@ -64,6 +67,33 @@ using AnimationEvents = Vector<Ref<AnimationEventBase>>;
+    }
+    static bool isDeletedValue(const WebCore::AnimatableProperty& value) {
+        return WTF::switchOn(value, [] (WebCore::CSSPropertyID property) { return HashTraits<WebCore::CSSPropertyID>::isDeletedValue(property); }
+                                  , [] (AtomString) { return false; });
```
Same comment about `const AtomString&` to avoid reference count churn.

graouts committed:
[web-animations] add an AnimatableProperty type to encapsulate either a standard or custom property
https://bugs.webkit.org/show_bug.cgi?id=249498

Reviewed by Darin Adler.

The animation code is currently heavily biased towards dealing with "standard" CSS properties, represented
by the CSSPropertyID enum. However, with the work to add interpolation support for custom properties (see
bug 249293) and to support CSS Transitions for custom properties (see bug 249399), we really need to deal
with both custom properties and standard properties everywhere we currently deal with only standard
properties.

In this patch we're adding a new type, AnimatableProperty, which is an std::variant<CSSPropertyID, AtomString>.
We currently use CSSPropertyID as a value for property lists which are HashSet<> and as keys to HashMap<>,
so we must add the relevant traits to support this use for AnimatableProperty.

* Source/WebCore/animation/WebAnimationTypes.h:
(WTF::DefaultHash<WebCore::AnimatableProperty>::hash):
(WTF::DefaultHash<WebCore::AnimatableProperty>::equal):
(WTF::HashTraits<WebCore::AnimatableProperty>::constructDeletedValue):
(WTF::HashTraits<WebCore::AnimatableProperty>::isDeletedValue):

Canonical link: https://commits.webkit.org/258040@main
